The Things Gateway CUPS

adriansmares opened the pull request:
<!--
Thanks for submitting a pull request. Please fill the template below,
otherwise we will not be able to process this pull request.
-->

**Summary:**
<!--
A short summary, referencing related issues:
Closes #0000, References #0000, etc.
-->

Closes #68.

**Changes:**
<!-- What are the changes made in this pull request? -->


- Add the `/api/v2/gateways/<gateway_id>` endpoint which renders the gateway information using the v2 format. The result is minimal and renders only the fields which are actually being used by The Things Gateway.
- Added `/api/v2/frequency-plans/<fp_id>` endpoint which the renders the frequency plan information using the v2 format.

**Notes for Reviewers:**
<!--
Motivate briefly why it is implemented this way, if that deviates from the
implementation proposal in the referenced issues.
- How should your reviewers approach this pull request?
- @mention reviewers with special requests or questions for them
-->


There is still an ongoing bug regarding the `frequency_plan_url` field, which I've been unable to correctly implement. The current implementation uses the `echo.Context.Request().Host` field in order to derive the host and the port of the current request, and derive the frequency plan URL. 

However, `Request().Host` is just an wrapper around the `Host` header sent by the HTTP client, and The Things Gateway sends only the host, without the port. As such, the URL generated does not contain the port, and the gateway cannot download the frequency plan if the IS is not listening on port 80. I've also attempted to use `Request().URL`, but again it does not contain the host / port.

One way to deal with this, which I'm not very fond of, is to add a setting for the `scheme://host:port` part of the generated URL.

**Release Notes: _(optional)_**
<!--
Any notes that we need to include in the Release Notes for the next release.
These notes are formatted as bullet points, written in past tense, and will be
combined with the labels of this Pull Request.
- Always mention changes in API, database models, configuration options or defaults.
- Are there any database migrations required?
- What are the functional or behavioral changes?
-->

- Added support for The Things Gateway.

adriansmares commented on cmd/ttn-lw-stack/commands/config.go:
```diff
@@ -43,7 +44,8 @@ type Config struct {
 	GCS              struct {
-		BasicStation cups.ServerConfig `name:"basic-station"`
+		BasicStation     cups.ServerConfig `name:"basic-station"`
+		TheThingsGateway ttg_cups.Config   `name:"the-things-gateway"`
```
Note: This configuration will instead be added to `gatewayconfigurationserver.Config` once the #435 is merged.

adriansmares commented on cmd/ttn-lw-stack/commands/start.go:
```diff
@@ -200,6 +200,9 @@ var (
 				cups := config.GCS.BasicStation.NewServer(c)
 				_ = cups
+
+				ttgCups := config.GCS.TheThingsGateway.NewServer(c)
```
Same as above.

htdvisser commented on cmd/ttn-lw-stack/commands/start.go:
```diff
@@ -200,6 +200,9 @@ var (
 				logger.Info("Setting up GCS")
```
Can we also expand GCS while we're at it?

htdvisser replied:
Oh this is already solved in #435

htdvisser commented on cmd/ttn-lw-stack/commands/config.go:
```diff
@@ -43,7 +44,8 @@ type Config struct {
 	Console          console.Config           `name:"console"`
 	GCS              struct {
-		BasicStation cups.ServerConfig `name:"basic-station"`
+		BasicStation     cups.ServerConfig `name:"basic-station"`
```
Maybe we should also alias this `cups` to `bs_cups` to avoid confusion

adriansmares replied:
Renamed in #435

htdvisser commented on pkg/thethingsgateway/cups/handlers.go:
```diff
@@ -0,0 +1,162 @@
+}
+
+func (s *Server) handleGatewayInfo(c echo.Context) error {
+	gatewayInfoFieldMask := types.FieldMask{Paths: []string{
```
No need to make this a variable if you only use it once

htdvisser commented on pkg/thethingsgateway/cups/server.go:
```diff
@@ -0,0 +1,135 @@
+)
+
+const (
+	defaultFirmwarePath  = "https://thethingsproducts.blob.core.windows.net/the-things-gateway/v1"
```
Default config values are specified in `cmd/internal`

htdvisser commented on pkg/thethingsgateway/cups/server.go:
```diff
@@ -0,0 +1,135 @@
+		WithConfig(conf),
+	}
+	if conf.Default.FirmwareURL == "" {
+		opts = append(opts, WithDefaultFirmwareURL(defaultFirmwarePath))
```
return error for invalid config

htdvisser commented on pkg/thethingsgateway/cups/server.go:
```diff
@@ -0,0 +1,135 @@
+		opts = append(opts, WithDefaultFirmwareURL(defaultFirmwarePath))
+	}
+	if conf.Default.UpdateChannel == "" {
+		opts = append(opts, WithDefaultUpdateChannel(defaultUpdateChannel))
```
return error for invalid config

htdvisser commented:
Let's have a look at the exact implementation of The Things Gateway. [RFC2616](https://tools.ietf.org/html/rfc2616#section-14.23) says that a Host header without port implies the default port, so if you tested on port 80/443 this is correct behavior. If not, let's create an issue in https://github.com/TheThingsProducts/gateway/issues

KrishnaIyer commented on pkg/thethingsgateway/cups/handlers.go:
```diff
@@ -0,0 +1,162 @@
+		"update_channel",
+		"auto_update",
+	}}
+	ctx := s.getContext(c)
```
inline this in the `Get` call.

adriansmares replied:
It is used twice.

KrishnaIyer commented on pkg/thethingsgateway/cups/server.go:
```diff
@@ -0,0 +1,135 @@
+	return ttnpb.NewGatewayRegistryClient(s.component.GetPeer(ctx, ttnpb.PeerInfo_ENTITY_REGISTRY, ids).Conn())
+}
+
+// Option configures the CUPSServer.
```
`CUPS` already includes server. Please apply to other instances.

adriansmares commented:
Fixed the requested changes and rebased on top of #435. I'll rebase it on top of `master` once #435 is merged.

> Let's have a look at the exact implementation of The Things Gateway. [RFC2616](https://tools.ietf.org/html/rfc2616#section-14.23) says that a Host header without port implies the default port, so if you tested on port 80/443 this is correct behavior. If not, let's create an issue in https://github.com/TheThingsProducts/gateway/issues

The testing was done using port `1885`, and it was not added to the `Host` header. Also as a general comment regarding tests - the latest tests were run while this was still an endpoint on the Identity Server, so some end-to-end tests should still be done before the merge. I'll see what I can do.

htdvisser commented on pkg/thethingsgateway/cups/handlers.go:
```diff
@@ -0,0 +1,161 @@
+	} else {
+		freqPlanURL := &url.URL{
+			Scheme: c.Scheme(),
+			Host:   c.Request().Host,
```
What we can do to work around https://github.com/TheThingsProducts/gateway/issues/60 is fill the stack's default ports for Scheme into Host in case the Host of the request doesn't contain a port and there is no indication of being behind a reverse proxy (perhaps by looking at the `X-Forwarded-For` header).

That way, if the stack is configured for 1885/8885, then the ports are added. If the stack is configured for 80/443 we don't break anything by unnecessarily adding the ports. I think we can just assume that the default http(s) ports are used if the stack is behind a reverse proxy.
